src/Microsoft.Data.SqlClient/netcore/src/Common/src/Microsoft/Data/Common/DbConnectionOptions.Common.cs

```diff
@@ -96,43 +96,18 @@ private static class SYNONYM
             internal const string UID = "uid";
         }
 
-        internal readonly bool HasPasswordKeyword;
-        internal readonly bool HasUserIdKeyword;
+        protected readonly string _usersConnectionString;
 
-        private readonly string _usersConnectionString;
-        private readonly Dictionary<string, string> _parsetable;
-        internal readonly NameValuePair _keyChain;
+        public abstract string UsersConnectionString(bool hidePassword);
 
-        internal Dictionary<string, string> Parsetable
-        {
-            get { return _parsetable; }
-        }
-
-        public string UsersConnectionString(bool hidePassword) =>
-            UsersConnectionString(hidePassword, false);
-
-        internal string UsersConnectionStringForTrace() => UsersConnectionString(true, true);
-
-        private string UsersConnectionString(bool hidePassword, bool forceHidePassword)
-        {
-            string connectionString = _usersConnectionString;
-            if (HasPasswordKeyword && (forceHidePassword || (hidePassword && !HasPersistablePassword)))
-            {
-                ReplacePasswordPwd(out connectionString, false);
-            }
-            return connectionString ?? string.Empty;
-        }
-
-        internal bool HasPersistablePassword => HasPasswordKeyword ?
-            ConvertValueToBoolean(KEY.Persist_Security_Info, false) :
-            true; // no password means persistable password so we don't have to munge
+        internal abstract string UsersConnectionStringForTrace();
+        public abstract bool IsEmpty { get;}
 
-        public bool ConvertValueToBoolean(string keyName, bool defaultValue)
+        protected static bool ConvertValueToBoolean(Dictionary<string, string> parsetable, string keyName, bool defaultValue)
```

Comment:
`protected` or `internal`?
Please update it to lowest scope possible. Thx!

Reply:
done